### PR TITLE
Voltaire voting - spec, swagger and types

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x] spec
- [x] swagger
- [x] `ApiConstructTransactionData` update
- [x] unit tests adjustment 

### Comments
We can OPTIONALLY vote now when construct tx. Here there are 3 options:
(a) abstain
(b) no confidence
(c) pass voting to representative

Representative can be defined via either its key hash or script hash.
Relation to delegation:
If `vote` field is omitted we always use ABSTAIN. If vote is chosen then as it is set. 
Voting can be realized alone OR together with Delegation.

More in spec.   

### Issue Number
adp-3212

### CODE REVIEW

File: specifications/api/swagger.yaml [hunk 1]
```diff
@@ -379,6 +379,26 @@ x-stakePoolId: &stakePoolId
...
+  type: string
+  format: bech32
+  example: drep_script1wqaz0q0zhtxlgn0ewssevn2mrtm30fgh2g7hr7z9rj5856457mm
+  description: DRep's key hash.
+
+x-drep: &drep
+  nullable: false
+  oneOf:
+    - <<: *drepKeyHash
+      title: DRep's key hash
```
Comment (Unisay) [inline specifications/api/swagger.yaml]: Maybe this title attribute could be moved to the corresponding `x-drep*Hash` close to other attributes?

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs [hunk 2]
```diff
@@ -2165,6 +2244,39 @@ instance ToJSON ApiStakeKeyIndex where
...
+            String $ encodeDRepKeyHashBech32 keyhash
+        DRepFromScriptHash scripthash ->
+            String $ encodeDRepScriptHashBech32 scripthash
+instance FromJSON ApiVoteAction where
+    parseJSON t =
+        parseAbstain t <|> parseNoConfidence t <|> parseKeyHash t <|> parseScriptHash t
+      where
+        parseKeyHash = withText "DRepKeyHash" $ \txt ->
+            case decodeDRepKeyHashBech32 txt of
+                Left (TextDecodingError err) -> fail err
```
Comment (Unisay) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types.hs:2261]:
```suggestion
                Left TextDecodingError err -> fail err
```

**Review by Unisay — APPROVED** [review]
👍🏼

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs [hunk 3]
```diff
@@ -1206,6 +1210,80 @@ newtype ApiEncryptMetadata = ApiEncryptMetadata
...
+    case fmap Bech32.dataPartToBytes <$> Bech32.decodeLenient t of
+        Left _ -> Left textDecodingError
+        Right (_, Just bytes) ->
+            Right $ DRepScriptHash bytes
+        Right _ -> Left textDecodingError
+      where
+        textDecodingError = TextDecodingError $ unwords
+            [ "Invalid DRep Script hash: expecting a Bech32 encoded value"
+            , "with human readable part of 'drep_script'."
+            ]
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]:
```suggestion
decodeDRepScriptHashBech32 :: T.Text -> Either TextDecodingError DRepScriptHash
decodeDRepScriptHashBech32 t =
    case fmap Bech32.dataPartToBytes <$> Bech32.decodeLenient t of
        Left _ -> Left textDecodingError
        Right (hrp', Just bytes)
            | hrp' == hrp -> Right $ DRepScriptHash bytes
        Right _ -> Left textDecodingError
      where
        textDecodingError = TextDecodingError $ unwords
            [ "Invalid DRep Script hash: expecting a Bech32 encoded value"
            , "with human readable part of 'drep_script'."
            ]
        hrp = [Bech32.humanReadablePart|drep_script|]
```

I would recommend to also check the human readable part — the error message suggests that it should be present anyway.

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs [hunk 4]
```diff
@@ -1206,6 +1210,80 @@ newtype ApiEncryptMetadata = ApiEncryptMetadata
...
+    case fmap Bech32.dataPartToBytes <$> Bech32.decodeLenient t of
+        Left _ -> Left textDecodingError
+        Right (_, Just bytes) ->
+            Right $ DRepKeyHash bytes
+        Right _ -> Left textDecodingError
+      where
+        textDecodingError = TextDecodingError $ unwords
+            [ "Invalid DRep key hash: expecting a Bech32 encoded value"
+            , "with human readable part of 'drep_vkh'."
+            ]
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]:
```suggestion
-- | Decode a Bech32 encoded 'DRepKeyHash'.
decodeDRepKeyHashBech32 :: T.Text -> Either TextDecodingError DRepKeyHash
decodeDRepKeyHashBech32 t =
    case fmap Bech32.dataPartToBytes <$> Bech32.decodeLenient t of
        Left _ -> Left textDecodingError
        Right (hrp', Just bytes) ->
            Right $ DRepKeyHash bytes
        Right _ -> Left textDecodingError
      where
        textDecodingError = TextDecodingError $ unwords
            [ "Invalid DRep key hash: expecting a Bech32 encoded value"
            , "with human readable part of 'drep_vkh'."
            ]
        hrp = [Bech32.humanReadablePart|drep_vkh|]
```

I would recommend to check the human-readable part as well — the error messages suggests it.

File: specifications/api/drep-voting.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
+# Specification: DRep voting
+
+This document specifies those aspects of the "Transactions New > Construct" HTTP endpoint that relate to DRep voting in Voltaire era.
```
Comment (HeinrichApfelmus) [inline specifications/api/drep-voting.md]:
```suggestion
This document specifies those aspects of the "Transactions New > Construct" HTTP endpoint that relate to DRep voting in the Voltaire era.
```

**Review by HeinrichApfelmus — APPROVED** [review]
Thank you! 😊 I would recommend to make the decoding of the Bech32 more strict, though — the error message suggests that anyway.

File: specifications/api/swagger.yaml (hunk 1 above)
Comment (paweljakubas) [reply specifications/api/swagger.yaml]: done

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]: good idea. done

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs (hunk 4 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]: good idea. done

File: specifications/api/drep-voting.md (hunk 5 above)
Comment (paweljakubas) [reply specifications/api/drep-voting.md]: done

**Pull request merged**